docs/strings.md

```diff
@@ -0,0 +1,529 @@
+
+The OpenStep specification and Java came at a time when the Unicode 1.0 specification was current and there appeared to be consensus that 16 bits was sufficient to represent any character in any language.
+This consensus changed in 1996 with the release of the Unicode 2.0 specification, expanding the encoding space to up to 1,112,064 code points.
+The current version of the Unicode specification, 13.0, uses just over 10% of this encoding space and even with the amount wasted on emoji it seems unlikely that the total space will be exhausted and so Verona is likely to be in a in a fortunate position of being able to define a string API without the definition of a character changing substantially over the lifetime of the language.
+
+It is incredibly unlikely; however, that a Verona string implementation will contain native Verona code that implements the full Unicode specification in an early release.
+The specification of the [Unicode collation algorithm](https://www.unicode.org/reports/tr10/) alone runs to several hundred pages of text.
+A string design will therefore need to be efficient when interoperating with a sandboxed C++ library, such as ICU.
+
+Who cares about strings?
+------------------------
+
+Verona aims to be an *infrastructure programming language*.
+This category encompasses target use cases such as device drivers, distributed databases, online translation engines, and so on.
+Some of these will not have any user-facing component and so will use strings as identifiers (for example, dictionary keys, path components, error indicators) without the need for any localisation.
+Others will use strings to present text to a human and will need rich unicode processing.
+Our goal should be to support both use cases with the same set of abstractions but avoid projects in the former category needing to pay any performance or code-size penalty relative to a simple array-of-probably-ASCII-bytes string representation.
+
+A lot of complex Verona systems are likely to include components that are at both extremes in terms of their use of strings.
+We should not require a global decision about how strings are represented, nor should we penalise modules that require only simple strings.
+
+Code units, code points, characters, and grapheme clusters
+----------------------------------------------------------
+
+Strings are, informally, sequences of characters.
+This is not always a useful abstraction as the definition of a character can, in different contexts, mean:
+
+ * A code unit, the smallest indivisible unit of a single or multi-byte encoding (e.g. `char` or `wchar_t` in C, `unichar` in Objective-C)
+ * A Unicode code point, which may be stored as 1-4 octets, depending on the encoding.
+ * A [grapheme cluster](https://www.unicode.org/reports/tr29/), a sequence of code points that are combined to display something that a human will consider to be a character.
+
+Iteration over strings is therefore very complex.
```

Comment:
Iteration is hard, but random access is much harder and users also expect that to just work. I'm not sure it's worth mentioning here.

Reply:
Agreed.  You can view random access on a string as partially cached iteration though (unless the string is ASCII, then the caching is free).